README.md
# audio

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-04-13
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-04-13
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/audio/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/audio/compare/0.1.0...0.1.1


src/audio/io.clj
(ns audio.io
  (require [clojure.java.io :refer [file output-stream input-stream]])
  (:import
    [java.io InputStream OutputStream]
    [java.nio ByteBuffer IntBuffer CharBuffer ShortBuffer LongBuffer]
    [javax.sound.sampled AudioFormat AudioFileFormat AudioFormat$Encoding
                                AudioFileFormat$Type AudioInputStream AudioSystem]
    [java.io ByteArrayInputStream]
           [java.nio ByteBuffer]
           (java.io RandomAccessFile FileInputStream FileOutputStream )

            ))


(def dir "/Users/michal/Desktop/bounces/")

(defn short-double
  "return value in range -1 to +1 as in matlab"
  [^Short sht]
  (double (/ sht 0x8000)))

(defn double-short [^double dbl]
  (short (* dbl 0x8000)))

(defn mult-ceiling [x ^double y]
  (let [result (* x y)]
    (cond (> result 0.9999999) 0.98
          (< result -0.9999999) -0.98
          :else result)))

(defn read-wav-data
  "loads wav file fname from sample and given length"
  ([fname length] (read-wav-data fname 0 length))
  ([fname from length]
  (let [file (java.io.RandomAccessFile. fname "r")
      channel (.getChannel file)
      fbuffer (.map channel java.nio.channels.FileChannel$MapMode/READ_ONLY (+ 44 (* from 4)) (+ 44 (* length 4) (* from 4)))
       _ (.order fbuffer java.nio.ByteOrder/LITTLE_ENDIAN)
        shortbuffer (.asShortBuffer fbuffer)
        _ (.close channel)
    Lseq  (for [x (range 0 (* length 2) 2)]
                   (.get shortbuffer x))
     Rseq  (for [x (range 1 (* length 2) 2)]
                   (.get shortbuffer x))]
     {:left (map short-double Lseq) :right (map short-double Rseq)})))


(defn interlave-samples [samples]
  (interleave (:left samples) (:right samples)))

(defn write-wav [fname samples fs bps channels]
  (let [header [18770 17990 8838 604 16727 17750 28006 8308 bps 0 1 channels fs 0 45328 2 4 16 24932 24948 9999 9999] ;last 4 bytes will be replaced later
        header-and-data (concat  header samples)
        size (* (count samples) (/ bps 8))
        total-size (+ 44 size) ; header is 44 bytes long
        file (java.io.RandomAccessFile. fname, "rw")
        channel (.getChannel file)
        buffer (.map channel java.nio.channels.FileChannel$MapMode/READ_WRITE 0 total-size)
    _ (.order buffer java.nio.ByteOrder/LITTLE_ENDIAN)
         fbuffer (.asShortBuffer buffer)]
    (.put fbuffer (short-array header-and-data) 0 (/ total-size 2))
    (.putLong buffer  40 size)
         (.close channel)))


(time (def audioData (read-wav-data  (str dir "01L.wav") 0 98784)))


;; (def fff (file (str dir "286.wav")))

(defn fill-data-buffer!
  [^ByteBuffer b-data data sample-bytes]
  (loop [idx 0
         data data]
    (when-not (empty? data)
      (let [b-idx (* sample-bytes idx)]
        (.putShort b-data b-idx (short (* (first data) Short/MAX_VALUE))))
      (recur (inc idx) (rest data))))

  b-data)

(defn *ceiling [^double x]
    (cond (> x 0.9999999) 0.999
          (< x -0.9999999) -0.999
          :else x))

(defn celing [xs]
  (map *ceiling xs))

(defn gain [^doubles xs gain]
  (map #(mult-ceiling gain %) xs))

(defn write-wav-data
  "Writes contents of data to a new wav file with path. Adds frame-rate and
  n-channels as file metadata for appropriate playback/consumption of the new
  audio file."
  ([samples path] (write-wav-data samples path 44100 2))
  ([samples path frame-rate n-channels]
  (let [interlaved (if (= (type samples) clojure.lang.PersistentArrayMap) (interlave-samples samples) samples)
        data (map *ceiling interlaved)
        frame-rate   (float frame-rate)
        n-channels   (int n-channels)
        sample-bytes (/ Short/SIZE 8)
        frame-bytes  (* sample-bytes n-channels)
        a-format     (AudioFormat. AudioFormat$Encoding/PCM_SIGNED
                                   frame-rate
                                   Short/SIZE
                                   n-channels
                                   frame-bytes
                                   frame-rate
                                   true)
        data-size    (count data)
        n-bytes      (* data-size sample-bytes)
        b-data       (ByteBuffer/allocate n-bytes)
        b-data       (fill-data-buffer! b-data data sample-bytes)
        stream       (AudioInputStream. (ByteArrayInputStream. (.array b-data))
                                        a-format
                                        data-size)
        f            (file path)
        f-type       AudioFileFormat$Type/WAVE]
    (AudioSystem/write stream f-type f))))

(defn wave-header [fname]
  (let [file (java.io.RandomAccessFile. fname "r")
      channel (.getChannel file)
      fileSize (.size channel)
      fbuffer (.map channel java.nio.channels.FileChannel$MapMode/READ_ONLY 0 44)
      _ (.order fbuffer java.nio.ByteOrder/LITTLE_ENDIAN)
        _ (.close channel)
        fs (.getInt fbuffer 24)
        size (.getInt fbuffer 40)
        channels (.getShort fbuffer 22)
        bps (.getShort fbuffer 34)
        samples (int (* (/ size bps channels) 8))]
  [fs samples channels bps fileSize]))

(defn fill-data-buffer!
  [^ByteBuffer b-data data sample-bytes]
  (loop [idx 0
         data data]
    (when-not (empty? data)
      (let [b-idx (* sample-bytes idx)]
        (.putShort b-data b-idx (short (* (first data) Short/MAX_VALUE))))
      (recur (inc idx) (rest data))))

  b-data)


(defn play-seq
  "Writes contents of data to a new wav file with path. Adds frame-rate and
  n-channels as file metadata for appropriate playback/consumption of the new
  audio file."
  ([inputData] (play-seq inputData 44100 2))
  ([inputData frame-rate n-channels]
  (let [data (-> (interlave-samples inputData) celing)
        frame-rate   (float frame-rate)
        n-channels   (int n-channels)
        sample-bytes (/ Short/SIZE 8)
        frame-bytes  (* sample-bytes n-channels)
        a-format     (AudioFormat. AudioFormat$Encoding/PCM_SIGNED
                                   frame-rate
                                   Short/SIZE
                                   n-channels
                                   frame-bytes
                                   frame-rate
                                   true)
        data-size    (count data)
        n-bytes      (* data-size sample-bytes)
        b-data       (ByteBuffer/allocate n-bytes)
        b-data       (fill-data-buffer! b-data data sample-bytes)
        stream       (AudioInputStream. (ByteArrayInputStream. (.array b-data))
                                        a-format
                                        data-size)
        clip         (AudioSystem/getClip)]
    (.open clip stream)
     (.start clip))))

(defn play-mono-seq
  "plays one channel at fs 44100, takes input data and option level gain"
  ([input] (play-mono-seq input 1))
  ([input level]
  (play-seq {:left (gain input level) :right (gain input level)})))

(defn toString
  "helper to writeSamplesToText"
  [data]
  (map str data))

(defn writeSamplesToText
  "write samples as text each line is sample represented as double between -1 and 1,
  if samples are stereo as {left: l right: r}, will write only left channel"
  [samples fname]
  (let [samplesToWrite (if (= (type samples) clojure.lang.PersistentArrayMap) (:left samples) samples)]
  (with-open [w (clojure.java.io/writer fname)]
  (doseq [line (toString samplesToWrite)]
    (.write w line)
    (.newLine w)))))

(defn getDataFromText
  "parse doubles from text fromated file"
  [fname]
  (with-open [rdr (clojure.java.io/reader fname)]
    (->> (reduce  conj [] (line-seq rdr))
         (map read-string))))


src/audio/try.clj
(ns audio.try
(:import (javax.sound.sampled AudioSystem DataLine$Info SourceDataLine
                              AudioFormat AudioFormat$Encoding)))

(def popular-format
  (AudioFormat. AudioFormat$Encoding/PCM_SIGNED
                48000 ; sample rate
                16    ; bits per sample
                2     ; channels
                4     ; frame size 2*16bits [bytes]
                48000 ; frame rate
                false ; little endian
                ))

(defn open-line [audio-format]
  (doto (AudioSystem/getLine (DataLine$Info. SourceDataLine audio-format))
    (.open audio-format)
    (.start)))

(defn sine [sample-rate freq]
  (let [term (/ 1 freq)
        samples (* term sample-rate)
        factor (/ (* 2 Math/PI) samples)]
    (map #(Math/sin (* % factor))
         (range samples))))

(defn amplitude [sample-size]
  (Math/pow 2 (- sample-size 1.1)))

(defn quantize [amplitude value]
  (int (* amplitude value)))

(defn unsigned-byte [x]
  (byte (if (> x 127) (- x 256) x)))

(defn little-endian [size x]
  (map #(-> (bit-shift-right x (* 8 %))
            (bit-and 255)
            unsigned-byte)
       (range size)))

(defn big-endian [size x]
  (reverse (little-endian size x)))

(defn sine-bytes [format freq]
  (let [{:keys [sampleSizeInBits frameSize
                sampleRate bigEndian]} (bean format)
        sample-size (/ sampleSizeInBits 8)
        ampl (amplitude sampleSizeInBits)
        tear (if bigEndian big-endian little-endian)]
    (->> (sine sampleRate freq)
         (map (partial quantize ampl))
         (map (partial tear sample-size))
         (map cycle)
         (map (partial take frameSize))
         (apply concat)
         byte-array)))

(defn recalc-data [{:keys [line] :as state} freq]
  (assoc state :data (sine-bytes (.getFormat line) freq)))

(defn play-data [{:keys [line data playing] :as state} agent]
  (when (and line data playing)
    (.write line data 0 (count data))
    (send-off agent play-data agent))
  state)

(defn pause [agent]
  (send agent assoc :playing false)
  (doto (:line @agent)
    .stop .flush))

(defn play [agent]
  (.start (:line @agent))
  (send agent assoc :playing true)
  (send-off agent play-data agent))

(defn change-freq [agent freq]
  (doto agent
    pause
    (send recalc-data freq)
    play))

(defn line-agent [line freq]
  (let [agent (agent {:line line})]
    (send agent recalc-data freq)
    (play agent)))

(defn tone-freq [x]
  (-> (Math/pow 2 (/ x 11)) (* 440) (/ 512)))

(def mountain-king
  [[0 1] [2 1] [3 1] [5 1]
   [7 1] [3 1] [7 2]
   [6 1] [2 1] [6 2]
   [5 1] [1 1] [5 2]
   [0 1] [2 1] [3 1] [5 1]
   [7 1] [3 1] [7 1] [12 1]
   [10 1] [7 1] [3 1] [7 1]
   [10 2] [-2 2]])

(defn play-melody [agent base-tone base-duration melody]
  (doseq [[tone duration] melody]
    (change-freq agent (tone-freq (+ base-tone tone)))
    (Thread/sleep (* base-duration duration)))
  (pause agent))

(def a (line-agent (open-line popular-format) 60))
(play-melody a  99 10 mountain-king)


src/audio/midi.clj
(ns audio.midi

  (:import (java.io File)
           (javax.sound.midi MidiSystem Sequence MidiMessage MidiEvent ShortMessage Track))
  (:require [overtone.live :as overtone]))



;; (piano)
;; (piano 67)
;; (Thread/sleep 256)
;; (piano 64)
;; (Thread/sleep 256)
;; (piano 60)

(defn ts [ms] (Thread/sleep ms))


(defn matmusic [f x y t]
  (loop [i 40]
    (if (< i 120)
      (do (play_piano i)
        (recur (int (f i x)))))))


(def c0 8.1757989156)
(def halfTone 1.05946309435941)
(def quaterTone 1.029302236643495)
(def b1 15.4338531643)
(def c0bound (* c0 quaterTone))

;; (matmusic * 1.03 0 120)

;; (doseq [note [60 64 67 72]] (piano note))

(def c0 8.1757989156)


(defn hz->midiTable []
  (loop [i 0 note c0 acc []]
    (if (> i 127) acc
      (recur (inc i) (* note halfTone) (conj acc {:midi i :hz note}))
      )))


(defn midi->hz [midiNote]
  (loop [note midiNote freq c0]
    (if (zero? note) freq
      (recur (dec note) (* halfTone freq)))))

(midi->hz 60)


(defn hz->midi
  "converts frequency to midi note"
  [herz]
  (loop [i 0 hz herz]
    (if (< hz c0bound) i
      (recur (inc i) (/ hz halfTone))
      )))

(hz->midi 17.3239144360)


(defn matmusic [fun t]
  (loop [i 40]
    (if (< i 120)
      (do (play_piano i)
        (recur (int (fun i 1.05)))))))



(def pokerface "/Users/michal/hits/2008_119_8_m_ladygaga_pokerface.mid")
(def m0 "/Users/michal/m0.mid")
(def midi5 "/Users/michal/61.mid")
(def midi6 "/Users/michal/62.mid")
(def midi7 "/Users/michal/63.mid")
(def midi8 "/Users/michal/60-126.mid")
(def midi9 "/Users/michal/60-120.mid")
(def smpl "/Users/michal/Desktop/bounces/Ralphi Rosario VS Bojan - You used to hold me Instrumental.wav")


(defn track->midi-events
  "takes midi track and returns vector of midi events"
  [track]
  (loop [i 0 events []]
    (if (= (inc i) (.size track)) events
      (recur (inc i) (conj events (.get track i))))))


(defn midiMessage
  "takes midi event and returns set of position of this event in ticks
  and midi note plus info if it's note on (0x90) or note off (0x80)"
  [event]
  (let [position (.getTick event)
        noteEvent (.getData1 (.getMessage event))
        command (.getCommand (.getMessage event))]
    {:pos position :note noteEvent :onOff command}))



(defn midi-parse
  "parses midi file, returns sequence of set: position, note.
  ignores note off, and velocity"
  ([file-name] (midi-parse file-name 0))
  ([file-name track]
   (let [note-on 0x90
         note-off 0x80
         sequence (MidiSystem/getSequence (File. file-name))
         track  (-> sequence .getTracks (aget track))]
     (->>
       track
       (track->midi-events)
       (filter #(instance? ShortMessage (.getMessage %)))
       (map midiMessage)

       (filter #(or (= (get % :onOff) 0x90) (= (get % :onOff) 0x80)))
;;        (map #(dissoc % :onOff))
       ))))




tt


(defn mapMidi [events]
  (loop [ev events acc {}]
    (if (empty? ev) acc
      (let [e (first ev)
            {:keys [note pos onOff]} e
            newEvent {:pos pos :onOff onOff}
            setNote (get acc note)
            newAcc (if (nil? setNote)
                     (assoc acc note [newEvent])
                     (assoc acc note(conj setNote newEvent)))]
      (recur (rest ev) newAcc)))))



(defn durations [noteEvents]
  (loop [notes (get noteEvents 1) acc []]
    (if (empty? notes) acc
      (let [pos1 (:pos (get notes 0))
            pos2 (:pos (get notes 1))
            duration (- pos2 pos1)]
        (recur (subvec notes 2) (conj  acc {:note (get noteEvents 0) :pos pos1 :dur duration}))))))


(defn parseMidiDurations [midiFile]
  (->> (midi-parse midiFile)
       (mapMidi)
       (map durations)
       (reduce #(into %1 %2))))









(defn playMidi [xs]
  (doseq [event xs]
    (play_piano (get event :notes) (get event :duration))))

;; (repeatedly 1 #(playMidi (events->seq tt)))



(def sampl (overtone/sample smpl))

(def sample1 (overtone/load-sample smpl))
(def dry (overtone/play-buf 1 sample1))


sample1
(overtone/defsynth lpf []
  (let [dry (overtone/play-buf 1 sample1)
        wet (overtone/lpf18 dry 300)]
      (overtone/out 1 [wet dry])))


(repeatedly 1 #(do (lpf)))





  ; (defn f1 [x]

;;   (do (println "started f1") (Thread/sleep 4000) (println "done f1 " x) x))

;; (defn f2 [y]
;;   (do (println "started f2") (Thread/sleep 5000) (println "done f2 " y) y))



;; (time (do (f1 1) (f2 2)))
;; (time (pmap #(%) [f1 f2]))
;; (time (pvalues (f1 1) (f2 2)))

;; (def letters (mapv (comp str char (partial + 65)) (range 26)))
;; (defn random-string [n]
;;   (apply str (take n (repeatedly #(rand-nth letters)))))

;; (defn rand-string [ll sl]
;;   (doall (take ll (repeatedly (partial random-string sl)))))



;; (def llist (rand-string 2000 10000))

;; (time (dorun (map clojure.string/lower-case llist)))

;; (time (dorun (map sort llist)))

;; (time (dorun (pmap sort llist)))

;; (time (doall (pmap (fn [x] (doall (pmap sort x))) (partition-all (/ (count llist) 16) llist))))

;; (defn multia
;;   ([x] (multia x "zero"))
;;   ([x y]
;;    (println x y)

;;    ))

;; (def v ["a" "b"])
;; (def vv [["a" "b"]["c" "d"]["e" "f"]["g" "h"]["i" "j"]])
;; (multia "a")
;; (multia "a" "b")
;; (apply multia v)

;;  (map (fn [x] (apply multia x)) vv)
;; (defn retry [x]
;;   (println "future available and the answer is:")
;;   )
;; (defn when-done [future-to-watch function-to-call]
;;           (future (function-to-call @future-to-watch)))

;;          (let [f (future (do (println "started ")(Thread/sleep 2000)) 42)]
;;             (when-done f #(retry %)))


src/audio/core.clj
(ns audio.core

  (require [clojure.java.io :refer [file output-stream input-stream]]
           [audio.io :as io])
  (:import
   [ java.net URL]
    [java.io ByteArrayInputStream]
    [java.io InputStream OutputStream]
    [javax.sound.sampled AudioFormat AudioFileFormat AudioFormat$Encoding
     Line Clip LineEvent LineListener LineEvent$Type TargetDataLine DataLine$Info
     SourceDataLine  Line$Info
     Port AudioFileFormat$Type AudioInputStream AudioSystem Mixer]
    [java.nio ByteBuffer IntBuffer CharBuffer ShortBuffer LongBuffer]
     (java.io RandomAccessFile FileInputStream FileOutputStream )))


(def dir "/Users/michal/Desktop/bounces/")

(defn short-double
  "return value in range -1 to +1 as in matlab"
  [^Short sht]
  (double (/ sht 0x8000)))

(defn double-short [^double dbl]
  (short (* dbl 0x8000)))

(defn mult-ceiling [^double x ^double y]
  (let [result (* x y)]
    (cond (> result 0.9999999) 0.98
          (< result -0.9999999) -0.98
          :else result)))

(defn gain [^doubles xs ^double gain]
  (map #(mult-ceiling gain %) xs))

(def audioData (io/read-wav-data  (str dir "0000.wav") 0 44100))
(def wnoise (io/read-wav-data  "/Users/michal/wnoise.wav" 0 4410000))
(def l (audioData :left))
(def lw (wnoise :left))

(defn sin [x] (Math/sin x))

(defn sinGenerator
  "generates 2channel mono seq {left: xs, right: xs} of sinusoid of given freq and length (ms) fs = 44100"
  ([freq length] (sinGenerator freq length 1))
  ([freq length amplitude]
  (let [fs 44100
        wavelengthInSamples (* 441/10 length)
        pi2 (* 2 (Math/PI))
        cycl (/ fs freq)
        piCycles (/ pi2 cycl)
        xs (map #(* amplitude (Math/sin (*  piCycles %))) (range wavelengthInSamples))]
    {:left xs, :right xs})))


(defn getR
  "helper function uses iterator i and order n"
  [^Integer i ^Integer n]
  (Math/sin (/ (* Math/PI (+ (* 2 i) 1)) (* 4 n))))


(deftype Coefficients [^Double ai ^Double d1i ^Double d2i])
(deftype Ws [^Double w0 ^Double w1 ^Double w2])

(defn coeffsLP
  "calclulate coefficients for sampling freq s, half power frequency f and order n"
  [^double f ^Integer n ^Integer s]
  (let [a (Math/tan (* (/ f s) Math/PI))
        a2 (* a a)
    coeffs (for [i (range n)]
      (let [r (getR i n)
            s (+ a2 (* 2.0 a r) 1)
            ai (/ a2 s)
            d1i (/ (* 2.0 (- 1.0 a2)) s)
            d2i (* -1.0 (/ (+ (- a2 (* 2.0 a r)) 1.0) s))]
        [ai d1i d2i]))]
    (vec coeffs)))

(type (get (get (coeffsLP 200.0 4 44000) 1) 2))

(defn coeffsHP
  "calclulate coefficients for sampling freq s, half power frequency f and order n"
  [^double f ^Integer n ^Integer s]
  (let [a (Math/tan (* (/ f s) Math/PI))
        a2 (* a a)
    coeffs (for [i (range n)]
      (let [r (getR i n)
            s (+ a2 (* 2.0 a r) 1)
            ai (/ 1.0 s) ; this is different from coeffsLP
            d1i (/ (* 2.0 (- 1.0 a2)) s)
            d2i (* -1.0 (/ (+ (- a2 (* 2.0 a r)) 1.0) s))]
        [ai d1i d2i]))]
    (vec coeffs)))

(defn coeffsBP
  "calclulate coefficients for sampling freq s, half power upper frequency f1, half power lower frequency f2 and order n"
  [^double f1 ^double f2 ^Integer n ^Integer s]
  (let [
        PIdivS (/ Math/PI s)
        a (/ (Math/cos (* PIdivS (+ f1 f2)))
        (Math/cos (* PIdivS (- f1 f2) )))
        ;a = cos(M_PI*(f1+f2)/s)/cos(M_PI*(f1-f2)/s);
        a2 (* a a)
        b (Math/tan (* PIdivS (- f1 f2)))
        b2 (* b b)
    coeffs (for [i (range n)]
      (let [r (getR i n)
            br (* b r)
            s (+ b2 (* 2.0 br) 1)
            ai (/ b2 s)
            d1i (/ (* 4.0 a (+ 1.0 br)) s)
            d2i (/ (* 2.0 (- b2 (* 2.0 a2) 1.0)) s)
            d3i (/ (* 4.0 a (- 1.0 br)) s)
            d4i (/ (- (+ (- b2 (* 2.0 br)) 1)) s)
            ]
        [ai d1i d2i d3i d4i]))]
    (vec coeffs)))

(defn getBPx
  "calculates filter transform over one sample x"
  [x coeffs w0i w1i w2i w3i w4i i n]
    (if (= i n) [x w0i w1i w2i w3i w4i]
      (let [[a d1 d2 d3 d4] (get coeffs i)
             w0 (get w0i i)
             w1 (get w1i i)
             w2 (get w2i i)
             w3 (get w3i i)
            w4 (get w4i i)
            w0new (+ (* d1 w1) (* d2 w2) (* d3 w3) (* d4 w4) x)
            w0i (assoc w0i i w0new)
            x (* a (+ (- w0new (* 2.0 w2)) w4))
            w4new w3
            w4i (assoc w4i i w4new)
            w3new w2
            w3i (assoc w3i i w3new)
            w2new w1
            w2i (assoc w2i i w2new)
           w1new w0new
            w1i (assoc w1i i w0new)
            ]
        (recur x coeffs w0i w1i w2i w3i w4i(inc i) n))))



(defn getLPx
  "calculates filter transform over one sample x"
  [x coeffs w0i w1i w2i i n]
    (if (= i n) [x w0i w1i w2i]
      (let [[^double a ^double d1 ^double d2] (get coeffs i)
             w0 (get w0i i)
             w1 (get w1i i)
             w2 (get w2i i)
            w0new (+ (* d1 w1) (* d2 w2) x)
            w0i (assoc w0i i w0new)
            x (* a (+ w0new (* 2.0 w1) w2))
            w2new w1
            w2i (assoc w2i i w2new)
           w1new w0new
            w1i (assoc w1i i w0new)
            ]
        (recur x coeffs w0i w1i w2i (inc i) n))))


(defn getHPx
  "calculates filter transform over one sample x"
  [x coeffs w0i w1i w2i i n]
    (if (= i n) [x w0i w1i w2i]
      (let [[^double a ^double d1 ^double d2] (get coeffs i)
             w0 (get w0i i)
             w1 (get w1i i)
             w2 (get w2i i)
            w0new (+ (* d1 w1) (* d2 w2) x)
            w0i (assoc w0i i w0new)
            x (* a (+ (- w0new (* 2.0 w1)) w2)) ; different from getLPx
            w2new w1
            w2i (assoc w2i i w2new)
           w1new w0new
            w1i (assoc w1i i w0new)
            ]
        (recur x coeffs w0i w1i w2i (inc i) n))))



(defn generateWs [n ws]
  (vec (for [x (range ws)] (vec (for [x (range n)] 0.0)))))

(defn butterLP_old
  "butterworth filter f - filter freq, n - order, s - sampling frequency"
  ([^doubles samples ^double f ^Integer n] (butterLP samples f n 44100))
  ([^doubles samples ^double f ^Integer n ^Integer s]
  (let [n (/ n 2)
        coefficients (coeffsLP f n s)
        [^doubles initW0 ^doubles initW1 ^doubles initW2] (generateWs n 3)
        ]
    (time (loop [xs samples acc [] w0 initW0 w1 initW1 w2 initW2 n n]
      (if (empty? xs) acc
        (let [
              [^double x ^doubles w0 ^doubles w1 ^doubles w2] (getLPx (first xs) coefficients w0 w1 w2 0 n)
              ]
          (recur (rest xs) (conj acc x) w0 w1 w2 n))))))))

(defn butterLP_1
  "butterworth filter f - filter freq, n - order, s - sampling frequency"
  ([^doubles samples ^double f ^Integer n] (butterLP samples f n 44100))
  ([^doubles samples ^double f ^Integer n ^Integer s]
  (let [n (/ n 2)
        coefficients (coeffsLP f n s)
        [^doubles initW0 ^doubles initW1 ^doubles initW2] (generateWs n 3)
        ]
   (time (loop [xs samples acc (transient []) w0 initW0 w1 initW1 w2 initW2 n n]
      (if (empty? xs) (persistent! acc)
        (let [
              [^double x ^doubles w0 ^doubles w1 ^doubles w2] (getLPx (first xs) coefficients w0 w1 w2 0 n)
              ]
          (recur (rest xs) (conj! acc x) w0 w1 w2 n))))))))

(defn butterLP
  "butterworth filter f - filter freq, n - order, s - sampling frequency"
  ([^doubles samples ^double f ^Integer n] (butterLP samples f n 44100))
  ([^doubles samples ^double f ^Integer n ^Integer s]
  (let [n (/ n 2)
        coefficients (coeffsLP f n s)
        [^doubles initW0 ^doubles initW1 ^doubles initW2] (generateWs n 3)
        ]
   (time (loop [xs samples acc (transient []) w0 initW0 w1 initW1 w2 initW2 n n]
      (if (empty? xs) (persistent! acc)
        (let [
              [^double x ^doubles w0 ^doubles w1 ^doubles w2] (getLPx (first xs) coefficients w0 w1 w2 0 n)
              ]
          (recur (rest xs) (conj! acc x) w0 w1 w2 n))))))))





(defn butterBP
  "butterworth filter f1 - upper filter freq,f2 - lower filter freq, n - order, s - sampling frequency"
  ([samples f1 f2 n] (butterBP samples f1 f2 n 44100))
  ([samples f1 f2 n s]
  (let [n (/ n 2)
        coefficients (coeffsBP f1 f2 n s)
        [initW0 initW1 initW2 initW3 initW4] (generateWs n 5)
        ]
    (loop [xs samples acc [] w0 initW0 w1 initW1 w2 initW2 w3 initW3 w4 initW4 n n]
      (if (empty? xs) acc
        (let [
              [x w0 w1 w2 w3 w4] (getBPx (first xs) coefficients w0 w1 w2 w3 w4 0 n)
              ]
          (recur (rest xs) (conj acc x) w0 w1 w2 w3 w4 n)))))))




(defn butterHP
  "butterworth filter f - filter freq, n - order, s - sampling frequency"
  ([^doubles samples ^double f ^Integer n] (butterHP samples f n 44100))
  ([^doubles samples ^double f ^Integer n ^Integer s]
  (let [n (/ n 2)
        coefficients (coeffsHP f n s) ; different from butterLP
        [^doubles initW0 ^doubles initW1 ^doubles initW2] (generateWs n 3)
        ]
    (loop [xs samples acc [] w0 initW0 w1 initW1 w2 initW2 n n]
      (if (empty? xs) acc
        (let [
              [^double x ^doubles w0 ^doubles w1 ^doubles w2] (getHPx (first xs) coefficients w0 w1 w2 0 n) ;; different from butterLP
              ]
          (recur (rest xs) (conj acc x) w0 w1 w2 n)))))))

(io/play-mono-seq (butterBP l 2800 1000 4) 0.8)
(butterLP lw 400 16 44100)
(butterHP (io/getDataFromText "/Users/michal/sines2.txt") 200 4 44100)
(io/writeSamplesToText lw "/Users/michal/lw.txt")

(defn bpf [s f b samples]
  (let [t (/ (* 2 Math/PI f) s)
        r (- 1 (/ (* 2 Math/PI b) s))
        twoSinT (* 2 (Math/sin t))
        r2 (* r r)
        twoRCosT (* 2 r (Math/cos (* 2 t)))
        A (/ (* (- 1 r) (Math/sqrt (+ (- 1 twoRCosT) r2))) twoSinT)
        t (* 2 r (Math/cos t))
        r (* r r)]
    (loop [xs samples acc [] a0 0 a1 0 a2 0]
      (if (empty? xs) acc
        (let [x (first xs)
              a0 (+ (- (* t a1) (* r a2)) x)
              a2 a1
              a1 a0]
          (recur (rest xs) (conj acc (* A (- a0 a2))) a0 a1 a2))))))
(io/play-mono-seq (bpf 44100 120 1 l) 8)
(io/play-mono-seq (butterHP l 400 16) 0.8)
(io/play-mono-seq (butterLP l 400 16) 0.8)
;; (io/play-mono-seq (io/getDataFromText "/Users/michal/sines2a.txt"))
;; (apply max (butterLP l 1000 14 44100))
;; (time (butterLP lw 1000 4))

;; (io/writeSamplesToText l "/Users/michal/l.txt")

;; (time (do (butterLP (:left wnoise )1000 4) (butterLP (:right wnoise )1000 4)))


;; different concurent solutions:
(defn LP [stereoSamples f n s]
  (let [l (:left stereoSamples)
        r (:right stereoSamples)
        lf (future (butterLP l f n s))
        rf (future (butterLP r f n s))
        lfdone @lf
        rfdone @rf]
    {:left lfdone, :right rfdone}
    ))

(time (:left (LP wnoise 1000 4 44100)))

(defn LP1 [stereoSamples f n s]
  (let [l (:left stereoSamples)
        r (:right stereoSamples)
        lrf (pmap (fn [x] (butterLP x f n s)) [l r])]

    {:left (first lrf) , :right (second lrf)}
    ))

;; (time (:left (LP1 wnoise 1000 4 44100)))

(defn LP2 [stereoSamples f n s]
  (let [l (:left stereoSamples)
        r (:right stereoSamples)
        lrf (pvalues (butterLP l f n s) (butterLP r f n s)) ]

    {:left (first lrf) , :right (second lrf)}
    ))
;; (io/play-seq (LP2 (io/read-wav-data "/Users/michal/Desktop/_refs2015/126 don diablo adair.wav" 88100) 200 16 44100))
(io/write-wav-data (LP2 wnoise 200 16 44100) "/Users/michal/wno200.wav")

(io/writeSamplesToText (io/read-wav-data "/Users/michal/Desktop/_refs2015/126 don diablo adair.wav" 44100) "/Users/michal/ddiablo.txt")
(io/writeSamplesToText (io/read-wav-data "/Users/michal/wnoise.wav" 882000) "/Users/michal/noise.txt")


;; (:left (io/read-wav-data "/Users/michal/Desktop/_refs2015/126 don diablo adair.wav" 881))
;; (io/play-mono-seq (io/getDataFromText "/Users/michal/sines2a.txt"))
(io/write-wav-data (io/getDataFromText "/Users/michal/sines2a .txt") "/Users/michal/cwno200.wav")


